### PR TITLE
Modify plugin, tests and doc

### PR DESCRIPTION
**Pull request by pnhofmann**
This is a relatively big diff - Sorry. Promises will be the last PR such big.

* Needed to clean up some things with plugin, that I thought were a good idea but where I am not happy any more (as explained in #367).
* Did a refactor of **plugin manager**
* After this refactor commands with 3 or even more words are possible without trouble
* Then fixed **tests** broken because above changes
* Added some code to jarviscli/tests/__init__.py to make testing easier: This code mocks everything of CmdInterpreter:JarvisAPI and helps verifying if corrent methods were called
* Finally updated **documentation:** Found this nice tool 'pydoc-markdown' to read python docstrings and format them as .md file

### CODE REVIEW

**Comment by pnhofmann** [comment]
Please take a look and tell me what you think.

**Comment by appi147** [comment]
Lots of work here!!

File: README.md [hunk 1]
```diff
@@ -30,28 +30,18 @@ You can start by typing `help` within the Jarvis command line to check what Jarv
...
 
-@plugin()
-def helloworld(jarvis, s):
-    """Prints \"hello world!\""""
-    jarvis.say("Hello World!")
 
-
-@plugin()
-def repeat(jarvis, s):
+@plugin("hellworld")
```
Comment (appi147) [inline README.md]:
```suggestion
@plugin("helloworld")
```

File: jarviscli/CmdInterpreter.py [hunk 2]
```diff
@@ -174,17 +180,6 @@ def __init__(self, first_reaction_text, prompt, directories=[], first_reaction=T
...
-                        "near",
-                        "pinpoint",
-                        "umbrella",
-                        {"update": ("location", "system")},
-                        "weather",
-                        ]
-
         self.fixed_responses = {"what time is it": "clock",
                                 "where am i": "pinpoint",
                                 "how are you": "how_are_you"
```
Comment (appi147) [inline jarviscli/CmdInterpreter.py]: Not related to your PR, but I think this has been removed

Comment (appi147) [reply jarviscli/CmdInterpreter.py]: Ok I found it in `plugin/ask.py` line 63.

But its not working if i dont give extra string to it

```
16:59 $ python jarviscli/
Jarvis' sound is by default disabled.
In order to let Jarvis talk out loud type: enable sound
Type 'help' for a list of available actions.
86 plugins loaded 4 plugins disabled. More information: status

~> Hi, what can I do for you?
how are you
I could not identify your command...
~> What can i do for you?
how are you jarvis
I am fine, How about you?
~> What can i do for you?
how are you 1
I am fine, How about you?
~> What can i do for you?


```

**Comment by appi147** [comment]
I really liked that you created test api and the documentations are really good and beginner friendly. Will look through other functions if they work.

Really good work !!

**Comment by appi147** [comment]
While we are cleaning, _config.yml and NEW_FUNCTIONALITIES.md are redundant. What you think?

**Comment by appi147** [comment]
Also, what do you think about aiml(ask.py). IMO it is a chatbot and can add million lines to it and it has fixed(random) answers. I dont think it serves any good.

**Comment by appi147** [comment]
I have a suggestion(in general), while we give `help` command, we see all the commands. 
```
Documented commands (type help <topic>):
========================================
ask        currencyconv  factor       increase  pinpoint  shutdown    umbrella 
battery    curve         file         ip        play      solve       update   
bye        curvesketch   gmail        limit     plot      speedtest   weather  
calc       decrease      go           lyrics    q         status      wiki     
calculate  dictionary    goodbye      match     quit      stopwatch   workspace
cancel     directory     hackathon    movie     quote     suspend   
check      disable       health       music     reboot    systeminfo
chuck      display       help         mute      remind    tell      
clear      enable        hibernate    near      roll      tempconv  
clock      equations     how          news      run       timer     
coin       evaluate      hybridsleep  open      say       todo      
cricket    exit          imgur        os        screen    translate 

```
But with commands like check, movie, etc. where there are sub-commands, it feels incomplete.

**Comment by appi147** [comment]
Related to PR:
maybe there's a problem with `cricket`

**Comment by appi147** [comment]
CURVE SKETCH:

Present
```
curve sketch y=sqrt((x+1)(x-1))

############### y=sqrt((x+1)(x-1)) ###############

##################### Graph ######################
Traceback (most recent call last):
  File "/usr/lib64/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib64/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/appi/Public/github/test/Jarvis/jarviscli/__main__.py", line 11, in <module>
    main()
  File "/home/appi/Public/github/test/Jarvis/jarviscli/__main__.py", line 7, in main
    jarvis.executor()
  File "/home/appi/Public/github/test/Jarvis/jarviscli/Jarvis.py", line 168, in executor
    self.cmdloop(self.first_reaction_text)
  File "/usr/lib64/python3.7/cmd.py", line 138, in cmdloop
    stop = self.onecmd(line)
  File "/usr/lib64/python3.7/cmd.py", line 217, in onecmd
    return func(arg)
  File "/home/appi/Public/github/test/Jarvis/jarviscli/plugin.py", line 209, in run
    self.get_plugins(sub_command).run(jarvis, s)
  File "/home/appi/Public/github/test/Jarvis/jarviscli/plugin.py", line 201, in run
    self._backend[0](jarvis.get_api(), s)
  File "/home/appi/Public/github/test/Jarvis/jarviscli/plugins/evaluator.py", line 275, in curvesketch
    plot().run(jarvis, s)
  File "/home/appi/Public/github/test/Jarvis/jarviscli/plugin.py", line 196, in run
    sub_command = jarvis.find_action(s, self.get_plugins().keys())
AttributeError: 'JarvisAPI' object has no attribute 'find_action'
```

Previous
```
curve sketch y=sqrt((x+1)(x-1))

############### y=sqrt((x+1)(x-1)) ###############

##################### Graph ######################


##################### Limit ######################
lim ->  ∞	= oo
lim -> -∞	= oo

############## Intersection x-axis ###############
0. x: -1
1. x: 1


############## Intersection y-axis ###############
1.0*I

##################### Factor #####################
((x - 1.0)*(x + 1.0))**0.5

################### Derivative ###################
1. Derivative: x/sqrt(x**2 - 1)
2. Derivative: -x**2/(x**2 - 1)**(3/2) + 1/sqrt(x**2 - 1)
3. Derivative: 3*x**3/(x**2 - 1)**(5/2) - 3*x/(x**2 - 1)**(3/2)

#################### Integral ####################
F(x) = x*sqrt(x**2 - 1)/2 - acosh(x)/2 + t

################ Maxima / Minima #################
(0/1.0*I) : unknown

################# Turning Point ##################

########################  ########################

```

File: jarviscli/plugins/converted.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+def check_weather(self, s):
+    """
+    weather in *: checks the current weather in any part of the globe.
+    -- Examples:
+        check weather in Canada
+    """
+    weatherIn.main(self._jarvis, s)
+
+
+@plugin("directory")
```
Comment (appi147) [inline jarviscli/plugins/converted.py]:
```suggestion
@plugin("direction")
```

**Comment by pnhofmann** [comment]
Big Thanks for reviewing!

> While we are cleaning, _config.yml and NEW_FUNCTIONALITIES.md are redundant. What you think?

Yes. Where was _config.yml even used? We should either maintain NEW_FUNCTIONALITIES.md or remove it.

> Also, what do you think about aiml(ask.py). IMO it is a chatbot and can add million lines to it and it has fixed(random) answers. I dont think it serves any good.

Honestly: I was mostly ignoring this part of the code ;). So yes, could live without it.

**Comment by appi147** [comment]
A general observation: jarvis crashes out now. Previous it would ask again "What i could do for you?" , but now terminal prompt comes.

**Comment by pnhofmann** [comment]
You are absolutely right - how could I not notice that?

**Comment by pnhofmann** [comment]
About crashes generally: Because we previously had crashes with e.g. no input or missing variable, too;
Just noticed, why not add a test, which simply calls **every** command with no and with a random/fixed string and check if something crashes?

**Comment by pnhofmann** [comment]
> maybe add information about autocomplete feature

Really like that idea! I just don't know if this is possible with GNU Readline.

Generally help functionality could be improved. Like better formatting or e.g. allow ``help check weather`` to print only help of "check weather".

**Comment by pnhofmann** [comment]
Cricket is still broken - but that's not related to my changes (any more).

**Comment by appi147** [comment]
> Cricket is still broken - but that's not related to my changes (any more).

Weird, it works in master.

I'll see what I can do

**Comment by appi147** [comment]
I think it can be merged will remove aiml/_config/NEW_F... in next PR

**Comment by pnhofmann** [comment]
Cricket works for you? That's strange...

```
~> Hi, what can I do for you?
cricket
Some error occurred, please open an issue on github!
Here is error:

Traceback (most recent call last):
  File "/home/phil/git/Jarvis/jarviscli/CmdInterpreter.py", line 155, in try_do
    do(self, s)
  File "/home/phil/git/Jarvis/jarviscli/plugin.py", line 219, in _plugin_run_with_network_error
    run_func(jarvis, s)
  File "/home/phil/git/Jarvis/jarviscli/plugin.py", line 201, in run
    self._backend[0](jarvis.get_api(), s)
  File "/home/phil/git/Jarvis/jarviscli/plugins/cricket.py", line 18, in __call__
    self.score(jarvis)
  File "/home/phil/git/Jarvis/jarviscli/plugins/cricket.py", line 109, in score
    matches = self.all_matches()
  File "/home/phil/git/Jarvis/jarviscli/plugins/cricket.py", line 32, in all_matches
    matches.append(match['mchdesc'])
KeyError: 'mchdesc'
```

**Comment by appi147** [comment]
Well now it fails for me. It worked when there were no matches

**Comment by appi147** [comment]
Well I found the cause of error. `pycricbuzz` was updated 3 weeks ago.
This [commit](https://github.com/codophobia/pycricbuzz/commit/ad29cedf88d782ac9596542096476403ac3b66b1) broke our package.

**Comment by pnhofmann** [comment]
This explains a lot. This raises the question if #352 was - contrary of what I thought, too - was a good idea.
At last it would probably a good idea to protect such code with test cases. I'll come up with a proposal later this week (-> #393).

Since this is not related to this PR, I'll merge anyway.

**Pull request merged**